tests/test_alembic.py

```diff
@@ -0,0 +1,169 @@
+
+
+def downgrade(alembic_config, migration):
+    subprocess.check_call(['alembic', 'downgrade', migration], cwd=path.dirname(alembic_config))
+
+
+def get_schema(session):
+    result = list(session.execute(text('''
+        SELECT type, name, tbl_name, sql
+        FROM sqlite_master
+        ORDER BY type, name, tbl_name
+        ''')))
+
+    return {(x[0], x[1], x[2]): x[3] for x in result}
+
+
+def assert_schemas_equal(left, right):
+    for (k, v) in left.items():
+        if k not in right:
+            raise AssertionError('Left contained {} but right did not'.format(k))
+        if not ddl_equal(v, right[k]):
+            raise AssertionError(
+                'Schema for {} did not match:\nLeft:\n{}\nRight:\n{}'
+                .format(k, v, right[k]))
+        right.pop(k)
+
+    if right:
+        raise AssertionError(
+            'Right had additional tables: {}'.format(right.keys()))
+
+
+def ddl_equal(left, right):
+    '''
+    Check the "tokenized" DDL is equivalent because, because sometimes Alembic schemas append
+    columns on the same line to the DDL comes out like:
+
+        column1 TEXT NOT NULL, column2 TEXT NOT NULL
+
+    and SQLAlchemy comes out:
+
+        column1 TEXT NOT NULL,
+        column2 TEXT NOT NULL
+    '''
+    # ignore the autoindex cases
+    if left is None and right is None:
+        return True
+
+    left = [x for x in WHITESPACE_REGEX.split(left) if x]
+    right = [x for x in WHITESPACE_REGEX.split(right) if x]
+
+    # Strip commas and quotes
+    left = [x.replace("\"", "").replace(",", "") for x in left]
+    right = [x.replace("\"", "").replace(",", "") for x in right]
+
+    return sorted(left) == sorted(right)
+
+
+def test_alembic_head_matches_db_models(tmpdir):
+    '''
+    This test is to make sure that our database models in `db.py` are always in sync with the schema
+    generated by `alembic upgrade head`.
+    '''
+    models_homedir = str(tmpdir.mkdir('models'))
+    subprocess.check_call(['sqlite3', os.path.join(models_homedir, 'svs.sqlite'), '.databases'])
+    engine = make_engine(models_homedir)
+    Base.metadata.create_all(bind=engine, checkfirst=False)
+    session = sessionmaker(engine)()
+    models_schema = get_schema(session)
+
+    alembic_homedir = str(tmpdir.mkdir('alembic'))
+    subprocess.check_call(['sqlite3', os.path.join(alembic_homedir, 'svs.sqlite'), '.databases'])
+    session = sessionmaker(make_engine(alembic_homedir))()
+    alembic_config = conftest._alembic_config(alembic_homedir)
+    upgrade(alembic_config, 'head')
+    alembic_schema = get_schema(session)
+
+    # The initial migration creates the table 'alembic_version', but this is
+    # not present in the schema created by `Base.metadata.create_all()`.
+    alembic_schema = {k: v for k, v in alembic_schema.items()
+                      if k[2] != 'alembic_version'}
+
+    assert_schemas_equal(alembic_schema, models_schema)
```

Comment:
so with regard to the issue you describe in this PR, I took a look at this today and I don't understand the following: why do check constraints appear at all in the generated alembic schema? alembic claims that they can't autogenerate check constraints (they've got an open issue for it: https://github.com/sqlalchemy/alembic/issues/508). do you understand why that is happening?

Reply:
I remember looking in to this some time in the past. If I remember right, the check constraints for SQLite booleans (`INT CHECK foo IN (0, 1)` type things) *do* work because they kinda have to otherwise the booleans just don't work at all. They are a special case that alembic *does* detect, but in general, check constraints that are explicitly defined are *not* detected.